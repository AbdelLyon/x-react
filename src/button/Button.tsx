import type { AnchorHTMLAttributes, ComponentType, ReactNode } from "react";
import { forwardRef } from "react";
import type { ButtonProps as ButtonRootProps } from "@nextui-org/button";
import { Button as ButtonRoot } from "@nextui-org/button";
import { cn } from "@/utils";

export interface ButtonProps extends ButtonRootProps {
  LinkComponent?: ComponentType<AnchorHTMLAttributes<HTMLAnchorElement>>;
  classNames?: {
    base?: string;
    beforeContent?: string;
    afterContent?: string;
    content?: string;
  };
}

export const Button = forwardRef<HTMLButtonElement, ButtonProps>(
  (
    {
      fullWidth = false,
      isLoading = false,
      isDisabled = false,
      startContent,
      endContent,
      className = "",
      LinkComponent,
      variant = "solid",
      classNames = {
        base: "",
        beforeContent: "",
        afterContent: "",
        content: "",
      },
      href,
      children,
      target,
      rel,
      ...props
    },
    ref,
  ) => {
    const baseStyles = cn(
      "transition-none font-normal border-1 rounded-md",
      fullWidth && "w-full",
      isLoading && "opacity-50 cursor-not-allowed",
      classNames.base,
      className,
    );

    const Content = (): ReactNode => (
      <>
        {startContent !== undefined && (
          <span className={cn("mr-2", classNames.beforeContent)}>
            {startContent}
          </span>
        )}
        <span className={classNames.content}>{children}</span>
        {endContent !== undefined && (
          <span className={cn("ml-2", classNames.afterContent)}>
            {endContent}
          </span>
        )}
      </>
    );

    const hasValidLink =
      href !== undefined &&
      href.length > 0 &&
      LinkComponent !== undefined &&
      LinkComponent !== undefined;

    if (hasValidLink) {
      return (
        <ButtonRoot
          ref={ref}
          {...props}
          as={LinkComponent}
          variant={variant}
          className={baseStyles}
          href={href}
          rel={target === "_blank" ? "noopener noreferrer" : rel}
          target={target}
        >
          <Content />
        </ButtonRoot>
      );
    }

    return (
      <ButtonRoot
        ref={ref}
        {...props}
        variant={variant}
        className={baseStyles}
        isDisabled={isDisabled}
      >
        <Content />
      </ButtonRoot>
    );
  },
);

Button.displayName = "Button";
