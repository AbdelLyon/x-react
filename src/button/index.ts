import "@/index.css";
export { Button, type ButtonProps } from "@/button/Button";
export { Buttons, type ButtonsProps } from "@/button/Buttons";
