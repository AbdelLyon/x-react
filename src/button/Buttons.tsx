import { forwardRef } from "react";
import type { ButtonGroupProps, ButtonProps } from "@nextui-org/react";
import { ButtonGroup } from "@nextui-org/react";
import { Button } from "./Button";

export type ButtonsProps = ButtonGroupProps & {
  buttons: Array<{
    key: string | number;
    label: React.ReactNode;
    buttonProps?: ButtonProps;
  }>;
};

export const Buttons = forwardRef(
  (
    { buttons, ...props }: ButtonsProps,
    ref: React.ForwardedRef<HTMLDivElement>,
  ) => {
    return (
      <ButtonGroup ref={ref} {...props}>
        {buttons.map(({ key, label, buttonProps }) => (
          <Button key={key} {...buttonProps}>
            {label}
          </Button>
        ))}
      </ButtonGroup>
    );
  },
);

Buttons.displayName = "Buttons";
