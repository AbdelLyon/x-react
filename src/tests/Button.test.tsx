import { Button } from "@/button";
import { describe, it, expect } from "vitest";
import { render } from "@testing-library/react";
import type { JSX } from "react";

describe("Button", () => {
  it("should match the snapshot with default props", () => {
    const { asFragment } = render(<Button>Click me</Button>);

    expect(asFragment()).toMatchSnapshot();
  });

  it("should match the snapshot with start content", () => {
    const { asFragment } = render(
      <Button startContent={<i>icon</i>}>Click me</Button>,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it("should match the snapshot with end content", () => {
    const { asFragment } = render(
      <Button endContent={<i>icon</i>}>Click me</Button>,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it("should match the snapshot with link props", () => {
    const MockLinkComponent = (
      props: React.AnchorHTMLAttributes<HTMLAnchorElement>,
    ): JSX.Element => <a {...props} />;
    const { asFragment } = render(
      <Button href="/test" LinkComponent={MockLinkComponent} target="_blank">
        Click me
      </Button>,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it("should match the snapshot with different variants", () => {
    const { asFragment } = render(
      <Button variant="bordered" fullWidth isDisabled>
        Click me
      </Button>,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it("should match the snapshot with custom styles", () => {
    const { asFragment } = render(
      <Button
        customStyles={{
          base: "custom-base",
          content: "custom-content",
        }}
      >
        Click me
      </Button>,
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
