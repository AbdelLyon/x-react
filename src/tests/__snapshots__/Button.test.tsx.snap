// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button > should match the snapshot with custom styles 1`] = `
<DocumentFragment>
  <button
    class="z-0 group relative inline-flex items-center justify-center box-border appearance-none select-none whitespace-nowrap subpixel-antialiased overflow-hidden tap-highlight-transparent data-[pressed=true]:scale-[0.97] outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 px-4 min-w-20 h-10 text-small gap-2 [&>svg]:max-w-[theme(spacing.8)] transition-transform-colors-opacity motion-reduce:transition-none bg-default text-default-foreground data-[hover=true]:opacity-hover transition-none font-normal border-1 rounded-md custom-base"
    type="button"
  >
    <span
      class="custom-content"
    >
      Click me
    </span>
  </button>
</DocumentFragment>
`;

exports[`Button > should match the snapshot with default props 1`] = `
<DocumentFragment>
  <button
    class="z-0 group relative inline-flex items-center justify-center box-border appearance-none select-none whitespace-nowrap subpixel-antialiased overflow-hidden tap-highlight-transparent data-[pressed=true]:scale-[0.97] outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 px-4 min-w-20 h-10 text-small gap-2 [&>svg]:max-w-[theme(spacing.8)] transition-transform-colors-opacity motion-reduce:transition-none bg-default text-default-foreground data-[hover=true]:opacity-hover transition-none font-normal border-1 rounded-md"
    type="button"
  >
    <span
      class=""
    >
      Click me
    </span>
  </button>
</DocumentFragment>
`;

exports[`Button > should match the snapshot with different variants 1`] = `
<DocumentFragment>
  <button
    class="z-0 group relative inline-flex items-center justify-center box-border appearance-none select-none whitespace-nowrap subpixel-antialiased overflow-hidden tap-highlight-transparent data-[pressed=true]:scale-[0.97] outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 px-4 min-w-20 h-10 text-small gap-2 opacity-disabled pointer-events-none [&>svg]:max-w-[theme(spacing.8)] transition-transform-colors-opacity motion-reduce:transition-none bg-transparent border-default text-foreground data-[hover=true]:opacity-hover transition-none font-normal border-1 rounded-md w-full"
    data-disabled="true"
    disabled=""
    type="button"
  >
    <span
      class=""
    >
      Click me
    </span>
  </button>
</DocumentFragment>
`;

exports[`Button > should match the snapshot with end content 1`] = `
<DocumentFragment>
  <button
    class="z-0 group relative inline-flex items-center justify-center box-border appearance-none select-none whitespace-nowrap subpixel-antialiased overflow-hidden tap-highlight-transparent data-[pressed=true]:scale-[0.97] outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 px-4 min-w-20 h-10 text-small gap-2 [&>svg]:max-w-[theme(spacing.8)] transition-transform-colors-opacity motion-reduce:transition-none bg-default text-default-foreground data-[hover=true]:opacity-hover transition-none font-normal border-1 rounded-md"
    type="button"
  >
    <span
      class=""
    >
      Click me
    </span>
    <span
      class="ml-2"
    >
      <i>
        icon
      </i>
    </span>
  </button>
</DocumentFragment>
`;

exports[`Button > should match the snapshot with link props 1`] = `
<DocumentFragment>
  <a
    class="z-0 group relative inline-flex items-center justify-center box-border appearance-none select-none whitespace-nowrap subpixel-antialiased overflow-hidden tap-highlight-transparent data-[pressed=true]:scale-[0.97] outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 px-4 min-w-20 h-10 text-small gap-2 [&>svg]:max-w-[theme(spacing.8)] transition-transform-colors-opacity motion-reduce:transition-none bg-default text-default-foreground data-[hover=true]:opacity-hover transition-none font-normal border-1 rounded-md"
    href="/test"
    rel="noopener noreferrer"
    role="button"
    tabindex="0"
    target="_blank"
  >
    <span
      class=""
    >
      Click me
    </span>
  </a>
</DocumentFragment>
`;

exports[`Button > should match the snapshot with start content 1`] = `
<DocumentFragment>
  <button
    class="z-0 group relative inline-flex items-center justify-center box-border appearance-none select-none whitespace-nowrap subpixel-antialiased overflow-hidden tap-highlight-transparent data-[pressed=true]:scale-[0.97] outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 px-4 min-w-20 h-10 text-small gap-2 [&>svg]:max-w-[theme(spacing.8)] transition-transform-colors-opacity motion-reduce:transition-none bg-default text-default-foreground data-[hover=true]:opacity-hover transition-none font-normal border-1 rounded-md"
    type="button"
  >
    <span
      class="mr-2"
    >
      <i>
        icon
      </i>
    </span>
    <span
      class=""
    >
      Click me
    </span>
  </button>
</DocumentFragment>
`;
